### PR TITLE
Add evaluation notebook

### PR DESCRIPTION
**Pull request by vasudevgupta7**
This PR add evaluation notebook on LibriSpeech dataset. Note: following discussion is for converted fine-tuned checkpoint.

@sayakpaul, when I am restricting sequence length (or padding) to some constant value (for using `tf.function()`), we are getting WER of 6% while without any padding, we are getting WER of 3%. There is a difference in performance in 2 cases as `original Wav2Vec2` doesn't accept any `attention_mask / padding_mask`.

Now if want to export to TFHub, we will have to restrict seqlen to some constant value (as SavedModel requires model signature to be constant) and we no longer will get 3% but we have no other option I guess.

Should I keep this notebook specific to TFHub export or specific to getting even better results (i.e 3% WER)?

@MorganR, please give your suggestions on this after you are back from vacations.

**Links to notebooks**
- Notebook with 3% WER: https://colab.research.google.com/github/vasudevgupta7/gsoc-wav2vec2/blob/notebook/notebooks/librispeech_evaluation_WER_3.ipynb
- Notebook with 6% WER: https://colab.research.google.com/github/vasudevgupta7/gsoc-wav2vec2/blob/notebook/notebooks/librispeech_evaluation_WER_6.ipynb

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks @vasudevgupta7.

> @sayakpaul, when I am restricting sequence length (or padding) to some constant value (for using tf.function()), we are getting WER of 6% while without any padding, we are getting WER of 3%. There is a difference in performance in 2 cases as original Wav2Vec2 doesn't accept any attention_mask / padding_mask.

I would suggest going with the two i.e. the eval with 6% WER and 3% WER. This will not only be helpful for the community but also will make the project even more transparent.

**Comment by vasudevgupta7** [comment]
Okay, will add another notebook also.

**Comment by vasudevgupta7** [comment]
@sayakpaul, @MorganR,

 Both notebooks are ready for review. Please give me your valuable feedback.

Notebook with 3% WER: https://colab.research.google.com/github/vasudevgupta7/gsoc-wav2vec2/blob/notebook/notebooks/librispeech_evaluation.ipynb

Notebook with 6% WER: https://colab.research.google.com/github/vasudevgupta7/gsoc-wav2vec2/blob/notebook/notebooks/librispeech_saved_model_evaluation.ipynb

Thanks!!

**Comment by sayakpaul** [comment]
@vasudevgupta7 thanks! 

I think the naming could be a bit better for the notebooks since they are a bit confusing to me. I'd indicate the WER in their naming for making the distinctions.

**Comment by sayakpaul** [comment]
More suggestions:

* I'd add necessary notes to indicate the parts that would be replaced with TF-Hub links later. 
* Instead of using native generators could we use `tf.data.Dataset.from_tensor_slices()` and then map the subsequent functions? 
* I'd keep the output small and clean (referring to the cell before you do WER). It might be a better user experience if we keep the following in the outputs: a. an audio widget of the original inputs, b. ground-truth labels, and c. predictions. If this limits the number of outputs we can show I am okay with that. Colab supports form elements like dropdown, sliders, etc. So if we want to use those to allow for better control over these little parameters, I am okay with that too. 

Let me know if anything is unclear.

**Comment by vasudevgupta7** [comment]
@sayakpaul thanks for the review and sorry for late reply.

> * Instead of using native generators could we use `tf.data.Dataset.from_tensor_slices()` and then map the subsequent functions?

This is hard to do as `tokenizer` doesn't work in `.map()`. I tried that but unable to get it working. I am using generators even in my main dataloading script ([see this](https://github.com/vasudevgupta7/gsoc-wav2vec2/blob/7b9169bf5e6d83471f6059e887ff1fc5822fb0a3/src/data_utils.py#L210)).

**Comment by vasudevgupta7** [comment]
> Did you try using `tf.py_function` or `tf._numpy_function`? Here's a relevant example: https://keras.io/examples/vision/randaugment/.

I have a small doubt: What's the benefit of using `tf.py_function` over generators as both have limited support currently? Generator code will look more cleaner right?

**Comment by sayakpaul** [comment]
With `tf.py_function` the mapping capabilities are possible which might be advantageous in different use-cases.

**Comment by vasudevgupta7** [comment]
> * Instead of using native generators could we use `tf.data.Dataset.from_tensor_slices()` and then map the subsequent functions?

@sayakpaul, for this model, we can't do that for the notebooks as `tf.data.Dataset.from_tensor_slices()` requires inputs to be of same dimensions (i.e all samples (both speech & text) should have same length to be able to use this) while we want our inputs to be of variable lengths.

Rest all suggestions are done.

**Comment by vasudevgupta7** [comment]
@MorganR, looking forward to feedback from you as well. I will merge once you approve.

Thanks!!

**Comment by vasudevgupta7** [comment]
Merging this PR as I need `main` branch to be updated for making final report. But would be happy to address any issues in separate PR if @MorganR has any suggestions on notebooks.
